Add the  exclude_flags argument to get_best_solution()

lukash opened the pull request:
Some preparatory/cleanup commits for the `libdnf::Solution` class followed by adding the new `exclude_flags` to it and using it for hawkey's `get_best_solution()`.

j-mracek commented:
@rh-atomic-bot try

j-mracek commented on libdnf/sack/Solution.cpp:
```diff
@@ -23,61 +23,56 @@
+                if (!query->empty()) {
                     nevra.reset(new Nevra(std::move(nevraObj)));
-                    query = std::move(queryCandidate);
-                    return true;
```
The return value was a performance improvement, because query.empty() is relatively costly request.

lukash replied:
Oh, I see, I haven't gone that deep. Still, the value wasn't used in the current code... probably better to keep it simple until it is needed?

j-mracek replied:
The code of _get_best_selector that is in development already use it or will use it. Please we have to think about design in advance. Or do you want to overwrite the code again and again?

lukash replied:
Where is the code? I didn't know about it. The design of the class seemed ad-hoc to me...

j-mracek replied:
https://github.com/rpm-software-management/libdnf/pull/472/files
in goal.cpp

j-mracek replied:
The solution class was created as a part of the PR, but it cannot be finished due to compatibility for dnf4.

lukash replied:
I see, thanks. I still think my changes make for a better design of the class. To solve the `empty()` issue, I would cache the emptiness value somewhere on the line of `Solution` - `Query` - `PackageSet`. The lower the better, but since both `Query` and `PackageSet` allow to access the underlying `Map` struct, we'd probably have to cache it in `Solution`. Ideally we get to a state when these classes don't need to leak their internal state and we can implement this optimization on the `PackageSet` level.

> Please we have to think about design in advance. Or do you want to overwrite the code again and again?

Iterating is inevitable to achieve quality code, especially if we want to turn it into a public API at some point. But I do agree thinking about the design carefully in advance saves the number of iterations...

lukash replied:
I've removed the commit that changed the `getBestSolution()` method to constructor.

j-mracek commented on libdnf/sack/Solution.cpp:
```diff
@@ -25,58 +25,56 @@ namespace libdnf {
-    baseQuery.apply();
-    std::unique_ptr<Query> queryCandidate(new Query(baseQuery));
+
+    query->addFilter(HY_PKG_EMPTY, HY_EQ, 1);
```
This is unneeded step. query == baseQuery therefore
query->addFilter(HY_PKG_EMPTY, HY_EQ, 1)
query->queryUnion(baseQuery);

result in same result like 
query->queryUnion(baseQuery);


Then line query->addFilter(HY_PKG_EMPTY, HY_EQ, 1) please move to line 78.

j-mracek replied:
The propose construct has a better performance because it allows to skip addFilter(HY_PKG_EMPTY, HY_EQ, 1) step when matches found.

lukash replied:
Fixed.

j-mracek commented on libdnf/sack/Solution.cpp:
```diff
@@ -25,58 +25,56 @@ namespace libdnf {
             return true;
         }
     }
-    queryCandidate->addFilter(HY_PKG_EMPTY, HY_EQ, 1);
```
Please add `query->addFilter(HY_PKG_EMPTY, HY_EQ, 1)`

j-mracek commented on python/hawkey/subject-py.cpp:
```diff
@@ -276,40 +278,47 @@ get_best_parser(_SubjectObject *self, PyObject *args, PyObject *kwds, HyNevra *n
-        &PyBool_Type, &with_filenames, &forms, &PyBool_Type, &with_src)) {
-        return NULL;
+        &PyBool_Type, &with_filenames, &forms, &PyBool_Type, &with_src, &exclude_flags_obj)) {
+        return std::make_pair(nullptr, std::unique_ptr<libdnf::Nevra>());
```
Please return NULL. NULL is required to raise an exception when parsing argument fails.

j-mracek commented on python/hawkey/subject-py.cpp:
```diff
@@ -276,40 +278,47 @@ get_best_parser(_SubjectObject *self, PyObject *args, PyObject *kwds, HyNevra *n
         cforms = fill_form<HyForm, _HY_FORM_STOP_>(forms);
         if (cforms.empty())
-            return NULL;
+            return std::make_pair(nullptr, std::unique_ptr<libdnf::Nevra>());
```
fill_form sets an error when cforms.empty(). To propagate it it must return NULL.

j-mracek commented on python/hawkey/subject-py.cpp:
```diff
@@ -266,8 +268,8 @@ nsvcap_possibilities(_SubjectObject *self, PyObject *args, PyObject *kwds)
-static PyObject *
-get_best_parser(_SubjectObject *self, PyObject *args, PyObject *kwds, HyNevra *nevra)
+static std::pair<PyObject*, std::unique_ptr<libdnf::Nevra>>
+get_solution(_SubjectObject *self, PyObject *args, PyObject *kwds)
```
Please what kind of issue you try to resolve by changes in get_best_parser? Return object static std::pair<PyObject*, std::unique_ptr<libdnf::Nevra>> is incorrect because it must allow to return NULL to propagate python exceptions.

lukash replied:
Unless I'm missing something, your point about returning NULL is not valid. This function is not called from Python, if it was, changing the return type would completely break it. There was, however, NULL returned from the `get_best_solution()` function on what used to be line 351, which I've removed, so I'll fix that.

This particular change in this commit was intended to be a greater simplification of the code, but I gradually encountered multiple reasons for why it needs to be the way it is, so what I ended up with is this minor improvement, where now the `unique_ptr` is returned instead of raw pointer, so that should be marginally safer, as well as the `unique_ptr` is now moved from the `Solution` object.

lukash commented:
I've moved the empty filter as requested and fixed the one place where I've removed returning `NULL` into Python. I think it should be good like this.

j-mracek commented:
Please refactor commit `Use libdnf::Subject directly in hawkey.Subject`.
Please keep only direct replacement of `hy_subject_get_best_solution` by `hawkey.Subject`. The rest of changes are not needed and it creates issues with no returning NULL to python when `PyErr_SetString()` is set.

j-mracek commented:
@rh-atomic-bot r+

j-mracek commented:
@rh-atomic-bot retry

j-mracek commented:
@rh-atomic-bot retry

j-mracek commented:
Following tests are failing `config-best.feature countme.feature errorlevel.feature install-fails.feature install-remove.feature module-install-package.feature`. @lukash Please could you investigate the issue?

lukash commented:
@j-mracek sorry about that, I made a mistake while rebasing after changes... I also had to rebase to latest master to make the tests pass.

@rh-atomic-bot retry

lukash commented:
@rh-atomic-bot retry

j-mracek commented:
@rh-atomic-bot retry

j-mracek commented:
@rh-atomic-bot retry

j-mracek commented:
@rh-atomic-bot r+

lukash commented:
@rh-atomic-bot retry
